eventserver/server_billable_handler.go

```diff
@@ -26,43 +30,120 @@ func BillableEventsHandler(store eventio.BillableEventReader, uaa auth.Authentic
 			return echo.NewHTTPError(http.StatusBadRequest, err)
 		}
 		// query the store
-		rows, err := store.GetBillableEventRows(filter)
+
+		rowOfRows := RowOfRows{}
+		defer rowOfRows.Close()
+
+		months, err := filter.SplitByMonth()
 		if err != nil {
 			return err
 		}
-		defer rows.Close()
-		// stream response to client
-		c.Response().Header().Set(echo.HeaderContentType, echo.MIMEApplicationJSONCharsetUTF8)
-		c.Response().WriteHeader(http.StatusOK)
-		if _, err := c.Response().Write([]byte("[\n")); err != nil {
-			return err
-		}
-		c.Response().Flush()
-		next := rows.Next()
-		delim := "\n"
-		for next {
-			b, err := rows.EventJSON()
+		for _, monthFilter := range months {
+			ok, err := consolidatedStore.IsRangeConsolidated(monthFilter)
 			if err != nil {
 				return err
 			}
-			if _, err := c.Response().Write(b); err != nil {
-				return err
-			}
-			next = rows.Next()
-			if next {
-				delim = ",\n"
+			var rows eventio.BillableEventRows
+			if ok {
+				rows, err = consolidatedStore.GetConsolidatedBillableEventRows(monthFilter)
 			} else {
-				delim = "\n"
+				rows, err = store.GetBillableEventRows(monthFilter)
 			}
-			if _, err := c.Response().Write([]byte(delim)); err != nil {
+			if err != nil {
 				return err
 			}
-			c.Response().Flush()
+			rowOfRows.RowsCollection = append(rowOfRows.RowsCollection, rows)
+		}
+
+
+		// stream response to client
+		c.Response().Header().Set(echo.HeaderContentType, echo.MIMEApplicationJSONCharsetUTF8)
+		c.Response().WriteHeader(http.StatusOK)
+		WriteRowsAsJson(c.Response(), c.Response(), &rowOfRows)
```

Comment:
A potential optimisation for the future: this block can be moved around slightly so the writing can be done in a more stream-like way if the reading of the rows into memory starts taking too much memory.

Reply:
Also: the `error` returned from `WriteRowsAsJson` is being silently swallowed here...

Reply:
> A potential optimisation for the future: I think this block can be moved around slightly so the writing can be done in a more stream-like way if the reading of the rows into memory starts taking too much memory.

But I do not think it is loading them in memory, is it? It opens all the queries to stream from DB at the beginning, but then it reads each one streaming them.

> Also: the error returned from WriteRowsAsJson is being silently swallowed here...

I will fix that